examples/text/plot_document_clustering.py

```diff
@@ -162,126 +46,187 @@ def is_interactive():
 
 
 # %%
-# Feature Extraction
-# ------------------
+# Define k-means pipeline
+# -----------------------
+# Here we define our k-means pipeline and the metrics to evaluate performance.
+#
+# Term-frequency inverse document-frequency (Tfidf) is used as a feature
+# extraction method.
+# TfidfVectorizer uses an in-memory vocabulary (a python dict) to map the most
+# frequent words to features indices and hence compute a word occurrence
+# frequency (sparse) matrix. The word frequencies are then reweighted using
+# the Inverse Document Frequency (IDF) vector collected feature-wise over
+# the corpus.
+#
+# Additionally, latent semantic analysis (LSA) is also be used to reduce
+# dimensionality and discover latent patterns in the data.
+#
+# Instead of ordinary k-means we use its more scalable cousin
+# minibatch k-means.
 
-labels = dataset.target
-true_k = np.unique(labels).shape[0]
+from sklearn.decomposition import TruncatedSVD
+from sklearn.feature_extraction.text import TfidfVectorizer
+from sklearn.pipeline import make_pipeline
+from sklearn.preprocessing import Normalizer
+from sklearn import metrics
+from sklearn.cluster import MiniBatchKMeans
+
+
+def kmeans_pipeline(
```

Comment:
Instead of this function, I would heavily rely on the notebook structure and add narrative.
In the end, we could create a cell for each of the following models:

```python
tf_model = make_pipeline(
    TfidfVectorizer(
        max_df=0.5,
        max_features=10000,
        min_df=2,
        stop_words="english",
        use_idf=False,
    ),
    KMeans(n_clusters=true_k, init="kmean++", random_state=0),
)

tf_idf_model = make_pipeline(
    TfidfVectorizer(
        max_df=0.5,
        max_features=10000,
        min_df=2,
        stop_words="english",
        use_idf=True,
    ),
    KMeans(n_clusters=true_k, init="kmean++", random_state=0),
)

tf_idf_lsa_model = make_pipeline(
    TfidfVectorizer(
        max_df=0.5,
        max_features=10000,
        min_df=2,
        stop_words="english",
        use_idf=True,
    ),
    TruncatedSVD(n_components=10, random_state=0),
    Normalizer(copy=False),
    KMeans(n_clusters=true_k, init="kmean++", random_state=0),
)
```

(The above models need to be double-checked, this is a draft of what I could observe from the code).
We could add a narration explaining what each pipeline is doing before creating it and finally make the comparison by evaluating each pipeline.